docs/zephyr_02_functional_requirements.sdoc

```diff
@@ -1317,6 +1317,248 @@ As a Zephyr OS user I want to... ???,
+UID: ZEP-134
+STATUS: Draft
+TYPE: Functional
+COMPONENT: Counting Semaphores
+TITLE: Counting Semaphores Exposes Ability to Reset Count to Zero
+STATEMENT: >>>
+When the Counting Semaphore reset is requested,
+
+Zephyr shall do all of the following:
+
+- set to zero the current count of the Counting Semaphore, and
+- abort all other outstanding attempts to interact with the Counting Semaphore.
+<<<
+
+[REQUIREMENT]
+UID: ZEP-135
+STATUS: Draft
+TYPE: Functional
+COMPONENT: Counting Semaphores
+TITLE: Counting Semaphore Reset Request Invalid Parameter is Catastrophic Error
+STATEMENT: >>>
+When the Counting Semaphore reset is requested,
+
+If the Counting Semaphore reset parameters are invalid,
+
+Zephyr shall report a catastrophic failure.
+<<<
+
+[REQUIREMENT]
+UID: ZEP-136
+STATUS: Draft
+TYPE: Functional
+COMPONENT: Counting Semaphores
+TITLE: Counting Semaphore Give unblocks pending Take by another thread
+STATEMENT: >>>
+While a thread is blocked waiting to Take from this Counting Semaphore,
+
+When the Counting Semaphore Give is requested,
+
+the Counting Semaphore shall do all of the following:
+
+- unblock the waiting thread indicating the Take request was successful; and
+- indicate to the Giving thread that the Give request was successful.
+<<<
+
+[REQUIREMENT]
+UID: ZEP-137
+STATUS: Draft
+TYPE: Functional
+COMPONENT: Counting Semaphores
+TITLE: Counting Semaphore Give without pending Take increases count by 1
+STATEMENT: >>>
+While no thread is blocked waiting to Take from this Counting Semaphore,
+
+When the Counting Semaphore Give is requested,
+
+the Counting Semaphore shall do all of the following:
+
+- increase the count of this semaphore by 1, up to the configured limit of this Semaphore; and
+- indicate to the Giving thread that the Give request was successful.
+<<<
+
+[REQUIREMENT]
+UID: ZEP-138
+STATUS: Draft
+TYPE: Functional
+COMPONENT: Counting Semaphores
+TITLE: Counting Semaphore Give Request Invalid Parameter is Catastrophic Error
+STATEMENT: >>>
+When the Counting Semaphore Give is requested,
+
+If the Counting Semaphore Give parameters are invalid,
+
+Zephyr shall report a catastrophic failure.
+<<<
+
+[REQUIREMENT]
+UID: ZEP-139
+STATUS: Draft
+TYPE: Functional
+COMPONENT: Counting Semaphores
+TITLE: Counting Semaphore Take when Count is >0
+STATEMENT: >>>
+While the Count of this Counting Semaphore is > 0,
+
+When the Counting Semaphore Take is requested,
+
+the Counting Semaphore shall do all of the following:
+
+- atomically reduce the Count of this Counting Semaphore by 1; and
+- indicate to the Taking thread that the Take request was successful.
+<<<
+
+[REQUIREMENT]
+UID: ZEP-140
+STATUS: Draft
+TYPE: Functional
+COMPONENT: Counting Semaphores
+TITLE: Counting Semaphore Take blocked when Count is 0
+STATEMENT: >>>
+While the Count of this Counting Semaphore is 0,
+
+When the Counting Semaphore Take is requested,
+
+the Counting Semaphore shall block the thread requesting a Counting Semaphore Take.
```

Comment:
**nitpick:** "... requesting this Counting Semaphore."

Reply:
Agreed. In other work I have used "said <item>". ;-) Either correction seems clear. Let's decide on one so we can be consistent.